### PR TITLE
STIX 2.1 export

### PR DESCRIPTION
**Pull request by jnwatson**

### CODE REVIEW

File: synapse/lib/stormlib/stix.py [hunk 1]
```diff
@@ -0,0 +1,818 @@
...
+        retn = {
+            'id': stixid,
+            'type': stixtype,
+            'spec_version': '2.1',
+        }
+        return retn
+
+    async def _addRel(self, srcid, reltype, targid):
+
+        stixid = f'relationship--{uuid4((srcid, targid))}'
```
Comment (jnwatson) [inline synapse/lib/stormlib/stix.py:778]: It isn't explicit what the cardinality of a particular relationship is, but it makes sense that you can have more than more relationship type between two stix nodes (like threat actor -> identity). I'd suggest adding reltype to the uuid4.

File: synapse/lib/stormlib/stix.py [hunk 2]
```diff
@@ -0,0 +1,818 @@
+import copy
```
Comment (jnwatson) [inline synapse/lib/stormlib/stix.py:1]: not used

File: synapse/lib/stormlib/stix.py [hunk 3]
```diff
@@ -0,0 +1,818 @@
...
+import json
+import uuid
+import asyncio
+import datetime
+import collections
```
Comment (jnwatson) [inline synapse/lib/stormlib/stix.py:6]: not used

File: synapse/lib/stormlib/stix.py [hunk 4]
```diff
@@ -0,0 +1,818 @@
...
+
+import stix2validator
```
Comment (jnwatson) [inline synapse/lib/stormlib/stix.py:8]: not used

File: synapse/lib/stormlib/stix.py [hunk 5]
```diff
@@ -0,0 +1,818 @@
...
+                            ),
+                        },
+                    },
+            }},
+
+        NOTE: The default config is an evolving set of mappings.  If you need to guarantee stable output please specify a config.
+
+        ''',
+         'type': {
+            'type': 'function', '_funcname': 'bundle',
```
Comment (jnwatson) [inline synapse/lib/stormlib/stix.py:554]: 3 space indent

File: synapse/tests/test_lib_stormlib_stix.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+                $bundle.add($node)
+
+                | spin | syn:tag=cno.mal.redtree $bundle.add($node, stixtype=malware)
+
+                fini { return($bundle) }
+            ''')
+
+            self.reqValidStix(bund)
+
+            #self.setTestBundle('basic.json', bund)
```
[the remark this answers is not in the document]
Comment (invisig0th) [reply synapse/tests/test_lib_stormlib_stix.py:141]: Leaving these as an example of how to update when the test changes.

File: synapse/tests/test_lib_stormlib_stix.py [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+                {[( it:exec:url=$execurl :exe=$file :url=http://vertex.link/ )]}
+
+                $bundle.add($node, stixtype=malware)
+
+                fini { return($bundle) }
+            ''', opts=opts)
+
+            self.reqValidStix(bund)
+
+            #self.setTestBundle('custom0.json', bund)
```
[the remark this answers is not in the document]
Comment (invisig0th) [reply synapse/tests/test_lib_stormlib_stix.py:173]: here too :D

File: synapse/lib/stormlib/stix.py (hunk 1 above)
Comment (invisig0th) [reply synapse/lib/stormlib/stix.py:778]: great catch.  thanks!

File: synapse/lib/ast.py [hunk 8]
```diff
@@ -3491,10 +3491,9 @@ async def run(self, runt, genr):
             raise s_stormctrl.StormReturn(valu)
 
         # no items in pipeline... execute
-        if self.kids:
+        if self.kids and self.isRuntSafe(runt):
```
Comment (vEpiphyte) [inline synapse/lib/ast.py:3494]: What is the return expected behavior if self.kids is true and self.isRuntSafe returns false?

File: synapse/tests/test_lib_stormlib_stix.py [hunk 9]
```diff
@@ -0,0 +1,263 @@
+import os
+import copy
+import json
+import aiohttp
```
Comment (vEpiphyte) [inline synapse/tests/test_lib_stormlib_stix.py:4]: add whitespace after stdlib imports

File: synapse/tests/test_lib_stormlib_stix.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vEpiphyte) [reply synapse/tests/test_lib_stormlib_stix.py:141]: adding that as a comment may be useful for the person who has to go and update the bundle.

File: synapse/lib/ast.py (hunk 8 above)
Comment (invisig0th) [reply synapse/lib/ast.py:3494]: no-op

**Review by invisig0th — APPROVED** [review]

**Review by vEpiphyte — APPROVED** [review]

**Pull request merged**